subsys/net/ip/net_tls_internal.h

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __NET_TLS_INTERNAL_H
+#define __NET_TLS_INTERNAL_H
+
+#include <net/net_context.h>
+#include <net/net_tls.h>
+
+#if defined(CONFIG_NET_TLS) || defined(CONFIG_NET_DTLS)
+
+#if defined(CONFIG_MBEDTLS)
+
+#if !defined(CONFIG_MBEDTLS_CFG_FILE)
+#include "mbedtls/config.h"
+#else
+#include CONFIG_MBEDTLS_CFG_FILE
+#endif /* CONFIG_MBEDTLS_CFG_FILE */
+
+#if defined(MBEDTLS_PLATFORM_C)
+#include "mbedtls/platform.h"
+#else
+#include <stdlib.h>
+#define mbedtls_time_t       time_t
+#define MBEDTLS_EXIT_SUCCESS EXIT_SUCCESS
+#define MBEDTLS_EXIT_FAILURE EXIT_FAILURE
+#endif /* MBEDTLS_PLATFORM_C */
+
+#include <mbedtls/ssl_cookie.h>
+#include <mbedtls/entropy.h>
+#include <mbedtls/ctr_drbg.h>
+#include <mbedtls/net_sockets.h>
+#include <mbedtls/x509.h>
+#include <mbedtls/x509_crt.h>
+#include <mbedtls/ssl.h>
+#include <mbedtls/error.h>
+#include <mbedtls/debug.h>
+
+#endif /* CONFIG_MBEDTLS */
+
+/** A list of secure tags that context should use */
+struct sec_tag_list {
+	/** An array of secure tags referencing TLS credentials */
+	sec_tag_t sec_tags[CONFIG_NET_MAX_CREDENTIALS_NUMBER];
+
+	/** Number of configured secure tags */
+	int sec_tag_count;
+};
+
+/** TLS context information. */
+struct net_tls {
+	/** Network context back pointer. */
+	struct net_context *context;
+
+	/** TLS specific option values. */
+	struct {
+		/** Select which credentials to use with TLS. */
+		struct sec_tag_list sec_tag_list;
+	} options;
+
+#if defined(CONFIG_MBEDTLS)
+	mbedtls_ssl_context ssl;
+	mbedtls_ssl_config config;
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+	mbedtls_x509_crt ca_chain;
+#endif /* MBEDTLS_X509_CRT_PARSE_C */
+#endif /* CONFIG_MBEDTLS */
+
+	/** Intermediated buffer to store decrypted content. */
+	char rx_ssl_buf[64];
+
+	/** Currently processed packet. */
+	struct net_pkt *rx_pkt;
+
+	/** Offset in the currently processed packet. */
+	int rx_offset;
+
+	/** TLS packet fifo. */
+	struct k_fifo rx_fifo;
+
+	/** Receive callback for TLS */
```

Comment:
Is this really "Receive callback for TLS"? Maybe it's "Original user recv callback for the context"?